### ISSUE TITLE
VCF file not bgzipp

### ISSUE THREAD

**Issue by ccastane9** [JimWhiting91/genotype_plot#8]
Hi, 

I am having a small issue, I have bgzipped and index my vcf file using the following code: 
bgzip -c my_data.vcf > my_data.vcf.gz
tabix -f -p vcf my_data.vcf.gz

However, whenever I try to use the file in genotype_plot, I received the error "vcf needs to be bgzipped"
Here is what I enter into R:
new_plot <- genotype_plot(vcf = file("ECA13_260_new.vcf.gz"), chr = 1, start = 11700000, end = 11800000, popmap = popmap, cluster = FALSE, snp_label_size = 10000, colour_scheme = c("#d4b9da","#e7298a","#980043"))

I do have bgzip properly installed as I've used this method before, so I'm wondering if I am putting the wrong information into the R code above. I made the popmap using data.frame().

Thank you for any help, 
Caitlin

**Comment by JimWhiting91** [comment]
Hi Caitlin,

Can you try modifying your code to
```
new_plot <- genotype_plot(vcf = "ECA13_260_new.vcf.gz", chr = 1, start = 11700000, end = 11800000, popmap = popmap, cluster = FALSE, snp_label_size = 10000, colour_scheme = c("#d4b9da","#e7298a","#980043"))
```

The basic functionality here is to parse the full path to the vcf file, although I can see how from the new example this maybe isn't clear. The error message you're seeing though is actually only looking for a .gz suffix, which suggests to me that the way you're parsing the vcf string within `file()` is what's causing that error to appear.

Cheers,
Jim

**Comment by ccastane9** [comment]
Thanks Jim, that did work and get the program moving! However, now it seems I have an issue with my parameters being "out of bounds". The chromosome I'm working on is roughly 43.7Mbp in size, so 11230000-11600000 should be within that range. Do you know what could be throwing the error? Below is what I entered into R and it's output -- thank you for helping me, I'm a bit of a novice at R. 

new_plot <- genotype_plot(vcf = "ECA13_260.vcf.gz", chr = 13, start = 11230000, end = 11700000, popmap = popmap, cluster = FALSE, snp_label_size = 10000, colour_scheme = c("#d4b9da","#e7298a","#980043"))
Scanning file to determine attributes.
File attributes:
meta lines: 4742
header_line: 4743
variant count: 9808
column count: 270
Meta line 4742 read in.
All meta lines processed
gt matrix initialized
Character matrix gt created.
Character matrix gt rows: 9808
Character matrix gt cols: 270
skip: 0
nrows: 9808
row_num: 0
Processed variant: 9808
All variants processed
Error in x@gt[i, j, drop = FALSE] : subscript out of bounds

**Comment by JimWhiting91** [comment]
So I think its read in the vcf ok, those initial lines come from the vcfR package reporting on the vcf, so you've got a vcf with 9808 snps (variant count: 9808). The error you're seeing is one that a few people have had whereby an individual in your popmap first column is not in your vcf (which is the "out of bounds" its referring to). So I'd just check your `popmap[,1]` output and check that all those individuals are definitely labelled the same in your vcf file (also watch out for any whitespace).

Sorry about this, this is the most common error message people are seeing and it is top of my list to update this with an error message that's actually clear!

**Comment by ccastane9** [comment]
Hi Jim, are there any other parameters I can check to try and resolve the error? I have double checked that all names and labeles match accordingly although I still receive the error. Maybe I'm not generating my popmap correctly (my commands are in bold)? I also tried to have my pop.table read in using read.table() and I received the same error. Many thanks for all of your help!

**> pop <- c("AKT003_BERN","AKT004_BERN","AKT006_BERN")
> indiv <- c("AKT003_BERN","AKT004_BERN","AKT006_BERN")
> testdf <- data.frame("indiv","pop")**
> new_plot <- genotype_plot(vcf = "ECA13_260.vcf.gz", chr = 13, start = 327, end = 1199, popmap = testdf, cluster = TRUE, snp_label_size = 10000, colour_scheme = c("#d4b9da","#e7298a","#980043"))
Scanning file to determine attributes.
File attributes:
  meta lines: 4742
  header_line: 4743
  variant count: 9
  column count: 270
Meta line 4742 read in.
All meta lines processed.
gt matrix initialized.
Character matrix gt created.
  Character matrix gt rows: 9
  Character matrix gt cols: 270
  skip: 0
  nrows: 9
  row_num: 0
Processed variant: 9
All variants processed
Error in x@gt[i, j, drop = FALSE] : subscript out of bounds

**Comment by JimWhiting91** [comment]
There shouldn't be any problem with reading the file in with read.table(), so if your popmap looks like this:
```
AKT003_BERN   AKT003_BERN
AKT004_BERN   AKT004_BERN
AKT006_BERN   AKT006_BERN
```
You should just be able read in with:
```
popmap <- read.table("path/to/popmap_file")
```

Otherwise you're right in that you can just make this manually in R, which is generally my preference! The issue with the code above is that you've got "" around the names of your vectors here: `data.frame("indiv","pop")`, which you can see below gives you a malformed popmap:
```
> pop <- c("AKT003_BERN","AKT004_BERN","AKT006_BERN")
> 
> indiv <- c("AKT003_BERN","AKT004_BERN","AKT006_BERN")
> testdf <- data.frame("indiv","pop")
> testdf
  X.indiv. X.pop.
1    indiv    pop
```

If you just remove those quotation marks you should be ok:
```
> pop <- c("AKT003_BERN","AKT004_BERN","AKT006_BERN")
> indiv <- c("AKT003_BERN","AKT004_BERN","AKT006_BERN")
> testdf <- data.frame(indiv,pop)
> testdf
        indiv         pop
1 AKT003_BERN AKT003_BERN
2 AKT004_BERN AKT004_BERN
3 AKT006_BERN AKT006_BERN
```

**Comment by ccastane9** [comment]
Unfortunately that still didn't work, whenever I extract information out of the VCF file, I get heading such as these:
AKT003_BERN.DP | AKT003_BERN.AD | AKT003_BERN.GT

Would that be an issue when ensuring my samples are input correctly? Would I need to add their extensions (####.g.vcf.gz)? 

Thanks for bearing with me, 
Caitlin

**Comment by JimWhiting91** [comment]
No you shouldn't need to add any extensions. To be honest it may be easiest if you can just attach a small subset of your vcf file with a few SNPs in so I can just check if there's any obvious reason why it's erroring out. Thanks!

**Comment by ccastane9** [comment]
Absolutely! I have sent a small set to: J.Whiting2@exeter.ac.uk

**Comment by JimWhiting91** [comment]
To anyone that comes across this, Caitlin sent me a subset of her VCF and I wasn't able to reproduce the error message, i.e. `Error in x@gt[i, j, drop = FALSE] : subscript out of bounds`. As far as I know, this error message can occur either because there is an individual in the popmap that is not in the VCF (which we troubleshooted here), or if R is out of date. The latter is most likely the problem here. Currently, updating to the most recent version of R and updating packages *should* resolve this issue.

I'm closing this issue currently, but will add a note once I've updated the package to avoid this particular error message regarding out-of-date installations of R as it is an awkward one to catch and an awkward one for people to resolve if using shared computing facilities.
